rfc/002-fault-tolerance/README.md

```diff
@@ -0,0 +1,487 @@
+
+## Implementation
+
+We briefly touch on the implementation of write and read path, before dedicating
+the majority of attention to the metastore
+
+### Write Path
+
+To achieve fault-tolerance on the write path, we need an acknowledgement
+mechanism to confirm whether events at the source have been successfully
+processed. The current architecture looks as follows:
+
+![Write Path](write-path.png)
+
+We are assuming that we can begin with reading from a reliable medium that
+supports an indexed stream of events, such as Apache Kafka. If no such data
+acquisition mechanism is available, we may consider a local disk buffer to
+emulate a reliable medium. For this proposal, working around an unreliable data
+transport channel is out of scope.
+
+A transaction that writes one or more partitions needs to store additional
+metadata, containing the offsets this partition is based on. This atomic commit
+of both the transaction and its source position allows the source to pick up
+exactly at the point of failure, without risk of re-ingesting twice or missing
+any events. It thus provides exactly-once semantics for our write path.
+
+In technical terms, we can extend the CAF Streaming message type to include a
+handle to the source and source specific position, and then trigger an ACK once
+the data has been persisted. VAST already implements a similar mechanism today
+that acknowledges when events arrived at an active partition, which (almost)
+guarantees that they are queryable.
+
+When an exising source reconnects to an index component after a fault, it
+needs to request the most recently acknowlegded source position from the
+catalog and advance to the returned position to prevent double ingestion.
+
+### Read Path
+
+For historical queries, we can achieve fault-tolerance on the read part by
+moving query execution into a dedicated process. This requires that the catalog
+is the sole owner of partitions, and that the catalog returns the set of
+candidate partitions relevant to answer a query.
+
+Live queries bypass the catalog and would therefore need to implement ACK'ing
+independently, at the granularity of table slices. We do not cover this process
+in detail, as it constitutes a subset of the ACK'ing for the write path outlined
+above.
+
+### Metastore
+
+The metastore primarily stores the state of partition data. We restrict our
+analysis to this use case. From the perspective of the catalog, the metastore
+API should expose itself as an actor that encapsulates the implementation
+details. This actor can rely on VAST's plugin architecture to implement the
+metastore-specific integration.
+
+#### Request-Response API
+
+The catalog requires the following interface from the metastore:
+
+1. `get: [uuid] → [partition]`
+2. `add: [partition] → ACK`
+3. `delete: [uuid] → ACK`
+4. `replace: [uuid], [partition] → ACK`
+
+The semantics of the mentioned types are as follows:
+
+- `partition`: metadata about a partition, such as its URL and its schema.
```

Comment:
This looks like it should be a superset of what is currently called the "partition info", i.e., the state that the catalog holds per partition except for the sparse index. Should we re-use that name here, as it already is established in the code base?